packages/ai-jsx/src/batteries/docs.tsx

```
@@ -1,154 +1,503 @@
+      return this.getStats();
+    }
+    this.loadingState = Corpus.LoadingState.IN_PROGRESS;
+    try {
+      await this.doCrawl();
+      this.loadingState = Corpus.LoadingState.COMPLETED;
+    } catch (e) {
+      this.loadingState = Corpus.LoadingState.FAILED;
+      throw e;
+    }
+    return this.getStats();
+  }
+
+  private async doCrawl(): Promise<void> {
+    pinoLogger.info('Starting crawl');
+    this.activePartitionsToToken.set(null, null);
+    while (this.activePartitionsToToken.size > 0) {
+      const partition = this.activePartitionsToToken.keys().next().value;
+      const token = this.activePartitionsToToken.get(partition);
+      this.activePartitionsToToken.delete(partition);
+
+      const response = await this.loader({ partition, token } as CorpusLoading.Request);
+      pinoLogger.info(
+        {
+          partitions: response.partitions?.length ?? 0,
+          documents: response.page?.documents.length ?? 0,
+          nextPage: response.page?.nextPageToken != undefined,
+        },
+        'Received response'
+      );
+      for (const newPartition of response.partitions ?? []) {
+        if (!this.completedPartitions.has(newPartition.name) && !this.activePartitionsToToken.has(newPartition.name)) {
+          this.activePartitionsToToken.set(newPartition.name, newPartition.firstPageToken ?? null);
+        }
+      }
+      if (response.page) {
+        this.documents.push(...response.page.documents);
+        pinoLogger.info('Starting vectorization');
+        const vectors = await this.vectorize(response.page.documents);
+        pinoLogger.info('Completed vectorization');
+        this.vectors.push(...vectors.flat());
+        if (response.page.nextPageToken) {
+          this.activePartitionsToToken.set(partition, response.page.nextPageToken);
+        } else {
+          this.completedPartitions.add(partition);
+        }
+      }
+      if (!response.page && !response.partitions) {
+        // Corner case - invalid response.
```

I'm slightly concerned about a seemingly-silent failure here. Is there any legit case where this would happen? If not, it seems better to throw.

---

Ok, but what about the former case? Is that a malformed response?

My only concern here is if the system is silently failing for a malformed response. If all the cases are valid responses, then what we have is fine.

---

I'm somewhat skeptical of treating an empty response and a non-empty response whose fields are empty as being different in general, but you've convinced me.  Updated.